backend/internal/th/ethereum.go

```diff
@@ -0,0 +1,224 @@
+	*bind.TransactOpts
+	PrivateKey *ecdsa.PrivateKey
+}
+
+const simulatedChainID = 1337
+
+var (
+	OneEther            = big.NewInt(params.Ether)
+	DefaultTokenBalance = big.NewInt(100)
+)
+
+func CreateEOA(t *testing.T) *EOA {
+	t.Helper()
+
+	priv, err := crypto.GenerateKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+	opts, err := bind.NewKeyedTransactorWithChainID(priv, big.NewInt(simulatedChainID))
+	if err != nil {
+		t.Fatal(err)
+	}
+	return &EOA{
+		PrivateKey:   priv,
+		TransactOpts: opts,
+	}
+}
+
+type BlockchainBackend struct {
+	*ethsimtracer.Backend
+	BankAccount *EOA
+	ChainID     int
+	Endpoint    string
+
+	rpc *rpc.Client
+}
+
+func NewBackend(t *testing.T, accounts ...common.Address) *BlockchainBackend {
+	t.Helper()
+
+	genesis := make(map[common.Address]types.Account)
+	for _, account := range accounts {
+		genesis[account] = types.Account{Balance: OneEther}
+	}
+	bankAccount := CreateEOA(t)
+	genesis[bankAccount.From] = types.Account{Balance: new(big.Int).Mul(big.NewInt(1000), OneEther)}
+
+	log.SetDefault(log.NewLogger(log.DiscardHandler()))
+	// TODO use random port
+	bk := ethsimtracer.NewBackend(genesis, func(nodeConf *node.Config, ethConf *ethconfig.Config) {
+		nodeConf.HTTPModules = append(nodeConf.HTTPModules, "debug", "eth")
+		nodeConf.HTTPHost = "127.0.0.1"
+		nodeConf.HTTPPort = 4242
+	})
+	endpoint := "http://127.0.0.1:4242"
+	client, err := rpc.Dial(endpoint)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return &BlockchainBackend{
+		Backend:     bk,
+		BankAccount: bankAccount,
+		ChainID:     simulatedChainID,
+		Endpoint:    endpoint,
+		rpc:         client,
+	}
+}
+
+func (b *BlockchainBackend) Client() *ethclient.Client {
+	return ethclient.NewClient(b.rpc)
+}
+
+func (b *BlockchainBackend) FundOneEther(t *testing.T, to common.Address) string {
+	t.Helper()
+
+	signedTx := b.MakeTx(t, b.BankAccount, &to, OneEther, nil)
+	if err := b.Client().SendTransaction(context.Background(), signedTx); err != nil {
+		t.Fatal(err)
+	}
+
+	b.Commit()
+	return signedTx.Hash().Hex()
+}
+
+func (b *BlockchainBackend) Fund(t *testing.T, to common.Address, amount *big.Int) string {
+	t.Helper()
+
+	signedTx := b.MakeTx(t, b.BankAccount, &to, amount, nil)
+	if err := b.Client().SendTransaction(context.Background(), signedTx); err != nil {
+		t.Fatal(err)
+	}
+	b.Commit()
+	return signedTx.Hash().Hex()
+}
+
+func (b *BlockchainBackend) MakeTx(t *testing.T, sender *EOA, to *common.Address, value *big.Int, data []byte) *types.Transaction {
+	t.Helper()
+
+	signedTx, err := types.SignTx(b.MakeUnsignedTx(t, sender.From, to, value, data), types.LatestSignerForChainID(big.NewInt(int64(b.ChainID))), sender.PrivateKey)
+	if err != nil {
+		t.Errorf("could not sign tx: %v", err)
+	}
+	return signedTx
+}
+
+func (b *BlockchainBackend) MakeUnsignedTx(t *testing.T, from common.Address, to *common.Address, value *big.Int, data []byte) *types.Transaction {
+	t.Helper()
+
+	head, _ := b.Backend.Client().HeaderByNumber(context.Background(), nil)
```

Comment:
```suggestion
	head, err := b.Backend.Client().HeaderByNumber(context.Background(), nil)
```
we should check for any errors coz we will be accessing the BaseFee value to calculate the gas price

Reply:
Since it's only a test file, I though that a simple panic was fine.